AIDevGallery/Samples/Open Source Models/Multimodal Models/DescribeImage.xaml.cs

```diff
@@ -153,68 +159,100 @@ private async IAsyncEnumerable<string> InferStreaming(string question, string im
         }
     }
 
-    private async Task LoadAndDescribeImage(StorageFile file)
+    private void CancelGeneration()
     {
-        ButtonTextBlock.Text = "Cancel";
-        ToolTipService.SetToolTip(SelectImageButton, "Cancel generation");
-        AutomationProperties.SetName(SelectImageButton, "Cancel generation");
-        Output.Text = string.Empty;
-        Loader.IsActive = true;
-        Loader.Visibility = Visibility.Visible;
+        _cts!.Cancel();
+        DescribeImageButton.IsEnabled = false;
+        ButtonTextBlock.Text = "Canceling, this could take a while";
+    }
 
-        _cts = new CancellationTokenSource();
+    private async void LoadImage(StorageFile file)
+    {
+        using IRandomAccessStream fileStream = await file.OpenAsync(FileAccessMode.Read);
+        BitmapImage bitmapImage = new();
+        await bitmapImage.SetSourceAsync(fileStream);
+        DefaultImage.Source = bitmapImage;
+        NarratorHelper.AnnounceImageChanged(DefaultImage, "Image changed: new upload"); // <exclude-line>
+        imageFile = file;
+    }
 
-        using (IRandomAccessStream fileStream = await file.OpenAsync(FileAccessMode.Read))
+    private async Task DescribeTheImage()
+    {
+        _cts = new CancellationTokenSource();
+        try
         {
-            BitmapImage bitmapImage = new();
-            await bitmapImage.SetSourceAsync(fileStream);
-            DefaultImage.Source = bitmapImage;
-            NarratorHelper.AnnounceImageChanged(DefaultImage, "Image changed: new upload"); // <exclude-line>
-        }
+            _cts.Token.ThrowIfCancellationRequested();
 
-        await Task.Run(async () =>
-        {
-            try
+            LoadImageButton.IsEnabled = false;
+            ButtonTextBlock.Text = "Cancel";
+            ToolTipService.SetToolTip(DescribeImageButton, "Cancel generation"); // <exclude-line>
+            AutomationProperties.SetName(DescribeImageButton, "Cancel generation"); // <exclude-line>
+            Output.Text = string.Empty;
+            Loader.IsActive = true;
+            Loader.Visibility = Visibility.Visible;
+
+            _cts.Token.ThrowIfCancellationRequested();
+
+            await Task.Run(async () =>
             {
-                await foreach (var part in InferStreaming("What is this image", file.Path, _cts.Token))
+                try
                 {
-                    DispatcherQueue.TryEnqueue(() => Output.Text += part);
+                    await foreach (var part in InferStreaming("What is this image", imageFile!.Path, _cts.Token))
+                    {
+                        DispatcherQueue.TryEnqueue(() => Output.Text += part);
+                    }
                 }
-            }
-            catch
-            {
-            }
-        });
-
-        ResetState();
+                catch
+                {
+                }
+                finally
+                {
+                    ResetState();
+                }
+            });
```

Comment:
You could use another Task to fully cancel it, if you want to. Something like this:
```cs
var task = Task.Run(... // Notice that this calls the Task.Run without awaiting it
await Task.WhenAny([task, Task.Delay(-1, tsc.Token)]); // -1 means infinite
```
That is basically it, but you need to be careful and handle exceptions properly.

Reply:
By "fully cancel it" I mean instantaneous, without having to wait for the internal inferencing to happen.

Reply:
This throws an unhandleable system exception from `ComputeLogits`. I tried to do this before